src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1106 @@
+ * 
+ * The driver should be configured() before use, and a Client should be set to
+ * enable a callback after a command is completed.
+ * Author:  Kevin Baichoo <kbaichoo@cs.stanford.edu>
+ * Date: 7/27/16
+ */
+
+use helpers::*;
+use core::{mem, slice};
+use core::cell::Cell;
+use common::take_cell::TakeCell;
+use pm;
+use nvic;
+
+//  These are the registers of the PicoCache -- a cache sole for the flash.
+#[allow(dead_code)]
+struct Picocache_Registers {
+    reserved_1:                             [u8;0x8],
+    picocache_control:                      usize,
+    picocache_status:                       usize,
+    reserved_2:                             [u8;0x10],
+    picocache_maintenance_register_0:       usize,
+    picocache_maintenance_register_1:       usize,
+    picocache_montior_configuration:        usize,
+    picocache_monitor_enable:               usize,
+    picocache_monitor_control:              usize,
+    picocache_monitor_status:               usize,
+    reserved_3:                             [u8;0xC4],
+    version:                                usize
+}
+
+//  Section 7 (the memory diagram) says the register starts at 0x400A0400. 
+const PICOCACHE_BASE_ADDRS : *mut Picocache_Registers = 0x400A0400 as *mut Picocache_Registers;
+
+//  Flush the cache. Should be called after every write!
+pub fn invalidate_cache() {
+    let registers : &mut Picocache_Registers = unsafe { 
+        mem::transmute(PICOCACHE_BASE_ADDRS)
+    };
+
+    volatile_store(&mut registers.picocache_maintenance_register_0, 0x1);
+}
+
+pub fn enable_picocache(enable : bool) {
+    let registers : &mut Picocache_Registers = unsafe { 
+        mem::transmute(PICOCACHE_BASE_ADDRS)
+    };
+    if (enable) {
+        volatile_store(&mut registers.picocache_control, 0x1);
+    }
+    else {
+        volatile_store(&mut registers.picocache_control, 0x0);
+    }
+
+}
+
+pub fn pico_enabled() -> bool {
+    let registers : &mut Picocache_Registers = unsafe { 
+        mem::transmute(PICOCACHE_BASE_ADDRS)
+    };
+   volatile_load(&registers.picocache_status) & 0x1 != 0
+}
+
+// Listing of the FLASHCALW register memory map.
+// Section 14.10 of the datasheet
+#[repr(C, packed)]
+#[allow(dead_code)]
+struct Registers {
+    control:                          usize,
+    command:                          usize,
+    status:                           usize,
+    parameter:                        usize,
+    version:                          usize,
+    general_purpose_fuse_register_hi: usize,
+    general_purpose_fuse_register_lo: usize,
+}
+
+const FLASHCALW_BASE_ADDRS : *mut Registers = 0x400A0000 as *mut Registers;
+
+#[allow(dead_code)]
+enum RegKey {
+    CONTROL,
+    COMMAND,
+    STATUS,
+    PARAMETER,
+    VERSION,
+    GPFRHI,
+    GPFRLO
+}
+
+/// ERROR codes
+pub enum Error {
+    CommandComplete,
+    LockE,
+    ProgE,
+    LockProgE,
+    ECC,
+}
+
+// High level commands to issue to the flash.
```

Comment:
Perhaps useful to document what the purpose of these commands are relative to `FlashCMD`